File: javascript/lasagna-master/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

export function cookingStatus(time) {
  if (time === 0) {
    return "Lasagna is done.";
  } else if (time > 0) {
    return "Not done, please wait.";
  } else {
    return "You forgot to set the timer.";
  }
}

export function preparationTime(layers, time = 2) {
    return layers.length * time;
}

export function quantities(ingredients) {
    return {
        noodles: ingredients.filter(i => i === 'noodles').length * 50,
        sauce: ingredients.filter(i => i === 'sauce').length * 0.2,
    }
}

export function addSecretIngredient(friendsList, myList) {
    myList.push(friendsList[friendsList.length - 1]);
}

export function scaleRecipe(recipe, portions) {
  const scaledRecipe = {};
  for (const ingredient in recipe) {
    scaledRecipe[ingredient] = recipe[ingredient] * portions / 2;
  }
  return scaledRecipe;
}

File: cpp/lasagna-master/lasagna_master.cpp
#include "lasagna_master.h"

namespace lasagna_master
{
    int preparationTime(std::vector<std::string> layers, int time)
    {
        return layers.size() * time;
    }

    amount quantities(std::vector<std::string> layers)
    {
        amount amnt;
        amnt.noodles = count(layers.begin(), layers.end(), "noodles") * 50;
        amnt.sauce = count(layers.begin(), layers.end(), "sauce") * 0.2;
        return amnt;
    }

    void addSecretIngredient(std::vector<std::string> &myList, const std::vector<std::string> &friendsList)
    {
        myList.back() = friendsList.back();
    }

    // Scales quantities based on the number of portions
    std::vector<double> scaleRecipe(std::vector<double> quantities, double num)
    {
        for (double &i : quantities)
        {
            i *= num / 2;
        }
        return quantities;
    }

    void addSecretIngredient(std::vector<std::string> &myList, const std::string secretIngredient)
    {
        myList.back() = secretIngredient;
    }
} // namespace lasagna_master
